crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,314 @@
+      remaining -= to_copy;
+      continue;  // continue to see if we've gotten |remaining| to 0
+    }
+    if (!ctx->ok) {
+      ctx->done = 1;  // Set EOF on cipher error.
+      break;
+    }
+    ctx->buf_len = 0;
+    ctx->buf_off = 0;
+    // |EVP_DecryptUpdate| may write up to cipher_block_size-1 more bytes than
+    // requested, so only read bytes we're sure we can decrypt in place.
+    int to_read = (int)sizeof(ctx->buf) - cipher_block_size + 1;
+    int bytes_read = BIO_read(next, ctx->buf, to_read);
+    if (bytes_read == 0 && BIO_eof(next)) {
+      // EVP_CipherFinal_ex may write up to one block to our buffer. If that
+      // happens, continue the loop to process the decrypted block as normal.
+      ctx->ok = EVP_CipherFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+      // We have more bytes to output, go back to top of loop.
+      if (ctx->buf_len > 0) {
+        continue;
+      }
+      ctx->done = 1;  // If we don't need to output any more bytes, set done.
+      return bytes_output;
+    } else if (bytes_read <= 0) {
+      BIO_copy_next_retry(b);
+      return bytes_output;
+    }
+    // Decrypt ciphertext in place, update |ctx->buf_len| with num bytes
+    // decrypted.
+    ctx->ok = EVP_DecryptUpdate(ctx->cipher, ctx->buf, &ctx->buf_len, ctx->buf,
+                                bytes_read);
+  }
+  return bytes_output;
+}
+
+static int enc_flush(BIO *b, BIO *next, BIO_ENC_CTX *ctx) {
+  GUARD_PTR(b);
+  GUARD_PTR(next);
+  GUARD_PTR(ctx);
+  while (ctx->buf_len > 0) {
+    int bytes_written = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (bytes_written <= 0) {
+      BIO_copy_next_retry(b);
+      return bytes_written;
+    }
+    ctx->buf_off += bytes_written;
+    ctx->buf_len -= bytes_written;
+    if (ctx->buf_len == 0 && !ctx->done) {
+      ctx->done = 1;
+      ctx->buf_off = 0;
+      ctx->ok = EVP_CipherFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+    }
+  }
+  return ctx->ok;
+}
+
+static int enc_write(BIO *b, const char *in, int inl) {
+  GUARD_PTR(b);
+  GUARD_PTR(in);
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL || !ctx->ok) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_consumed = 0;
+  int remaining = inl;
+  const int max_crypt_size =
+      (int)sizeof(ctx->buf) - EVP_CIPHER_CTX_block_size(ctx->cipher) + 1;
+  while (remaining > 0) {
+    assert(bytes_consumed + remaining == inl);
+    if (ctx->buf_len == 0) {
+      ctx->buf_off = 0;
+      int to_encrypt = remaining <= max_crypt_size ? remaining : max_crypt_size;
+      ctx->ok = EVP_EncryptUpdate(ctx->cipher, ctx->buf, &ctx->buf_len,
+                                  (uint8_t *)in, to_encrypt);
+      if (!ctx->ok) {
+        goto out;
+      };
+      bytes_consumed += to_encrypt;
+      remaining -= to_encrypt;
+    }
+    int bytes_written = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (bytes_written <= 0) {
```

Comment:
For [`BIO_write`](https://github.com/aws/aws-lc/blob/0ea1cb1ac4b283b318be8e87bc02022c1215be68/include/openssl/bio.h#L138-L143), I think we should distinguish `bytes_written == 0` (no bytes written) from `bytes_written < 0` (an error). On an error, we might need to set `ctx->ok = 0` and `ctx->done = 1`?

Reply:
for same reason as above, i think we should copy underlying BIO's retry state and return without setting `ctx->ok` or `ctx->done` -- there might still be a recoverable case.

but we should probably do a better job of surfacing the underlying return code (right now we just `return bytes_consumed`, which may be >= 0, but now < 0).

how about this for handling <= 0 return value from BIO_read/BIO_write? if `bytes_consumed` > 0, return that so caller knows how much to increment the input pointer on next call. else, return return value from BIO_read/BIO_write.